services/code_hosting/git_hub_service.rb

```diff
@@ -1,131 +1,409 @@
+            end
+          end
+          fastfile_json_path = File.join(path, "fastfile.json")
+          FileUtils.touch(fastfile_json_path) unless File.exist?(fastfile_json_path) && !File.zero?(fastfile_json_path)
+          File.write(fastfile_json_path, JSON.pretty_generate(fastfile_config))
+          self.cache[[repo, branch].join("/")] = fastfile, fastfile_config
+          return fastfile, fastfile_config
+        rescue ArgumentError
+          self.clone(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential,
+            path: path
+          )
+          fastfile, fastfile_config = self.peek_fastfile_configuration(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential,
+            path: path,
+            cache: cache
+          )
+          return fastfile, fastfile_config
+        rescue RuntimeError
+          # This is because no Fastfile config was found, so we cannot go further.
+          return {}
+        end
+      end
+
+      # Class method that shallow clones the repo on the target branch.
+      # @param [String] repo_url
+      # @param [String] branch
+      # @param [GithubProviderCredential] provider_credential
+      # @return [Git::Base]
+      def clone(repo_url: nil, name: nil, branch: nil, sha: nil, provider_credential: nil, path: nil)
+        repo = self.repo_from_url(repo_url)
+
+        folder_name = name || repo.split("/").last
+        clone_path = path || self.temp_path
 
-    GitHubService.status_context_prefix = "fastlane.ci: "
+        FileUtils.rm_rf(File.join(clone_path, folder_name)) if File.directory?(File.join(clone_path, folder_name))
+        FileUtils.mkdir_p(clone_path) unless File.directory?(clone_path)
+
+        self.setup_auth(repo_url: repo_url, provider_credential: provider_credential, path: clone_path)
+
+        if sha.nil?
+          Git.clone(url_from_repo(repo_url), folder_name,
+                    path: clone_path,
+                    recursive: true,
+                    depth: 1)
+        else
+          Git.clone(url_from_repo(repo_url), folder_name,
+                    path: clone_path,
+                    recursive: true)
+        end
+
+        git = Git.open(File.join(clone_path, folder_name))
+        git.branch(branch).checkout if sha.nil? && !branch.nil?
+        git.reset_hard(git.gcommit(sha)) unless sha.nil?
+
+        self.unset_auth
+
+        return git
+      end
+
+      # Class method that finds the directory for the first Fastfile found given a root_path
+      # @param [String] root_path
+      # @return [String, nil] the path of the Fastfile or nil if no Fastfile found.
+      def fastfile_path(root_path: nil)
+        fastfiles = Dir[File.join(root_path, "fastlane/Fastfile")]
+        fastfiles = Dir[File.join(root_path, "**/fastlane/Fastfile")] if fastfiles.count == 0
+        fastfile_path = fastfiles&.first
```

Comment:
what is the `&` for here?

Reply:
Yeah, doesn't makes sense, `fastfiles` will never be `nil` because of `Dir[]`, so `first` will be the `nil` returner, does not safe us from nothing. Will fix it.